src/payment/strategies/googlepay/googlepay-payment-strategy.ts

```diff
@@ -134,81 +144,50 @@ export default class GooglePayPaymentStrategy implements PaymentStrategy {
             return options.googlepaystripe;
         }
 
-        throw new InvalidArgumentError();
+        throw new InvalidArgumentError('Unable to initialize payment because "options.googlepay" argument is not provided.');
     }
 
-    private _getPayment(): PaymentMethodData {
-        if (!this._methodId) {
+    private async _getPayment(methodId: string): Promise<PaymentMethodData> {
+        if (!methodId) {
             throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
         }
 
-        const state = this._store.getState();
-        const paymentMethod = state.paymentMethods.getPaymentMethod(this._methodId);
-
-        if (!paymentMethod) {
-            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
-        }
-
-        if (!paymentMethod.initializationData.nonce) {
-            throw new MissingDataError(MissingDataErrorType.MissingPayment);
-        }
+        let state = this._store.getState();
+        this._paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
 
         let nonce;
 
-        if (this._methodId === 'googlepayadyenv2') {
+        if (methodId === 'googlepayadyenv2') {
             nonce = JSON.stringify({
                 type: AdyenPaymentMethodType.GooglePay,
-                googlePayToken: paymentMethod.initializationData.nonce,
+                googlePayToken: this._paymentMethod.initializationData.nonce,
                 browser_info: getBrowserInfo(),
             });
         } else {
-            nonce = paymentMethod.initializationData.nonce;
+            nonce = this._paymentMethod.initializationData.nonce;
         }
-
         const paymentData = {
-            method: this._methodId,
+            method: methodId,
             nonce,
-            cardInformation: paymentMethod.initializationData.card_information,
+            cardInformation: this._paymentMethod.initializationData.card_information,
         };
 
+        if (this._paymentMethod.initializationData.nonce) {
```

Comment:
shouldn't you be checking `nonce` rather than `this._paymentMethod.initializationData.nonce`

Reply:
no, because in the case` if (methodId === 'googlepayadyenv2')` the `nonce` value couldn't be empty then it isn't going to call the backend to refresh the value, if `this._paymentMethod.initializationData.nonce` is empty

Reply:
I think this is quite confusing, you have `this._paymentMethod.initializationData.nonce` and `nonce` and in some cases they might be different things.

To make you can refactor this logic so it's easier to read by removing ambiguity around what `nonce` actually holds.

```js
const { nonce } = this._paymentMethod.initializationData;

if (nonce) {
  state = await this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId));
  this._paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
}

const { card_information: cardInformation } = this._paymentMethod.initializationData;

return {
  methodId,
  paymentData: {
     method: methodId,
     cardInformation,
     nonce: this._getNonce(this._paymentMethod),
  },
};
```
```js
/* omitted types for simplicty */
function _getNonce({ methodId, initializationData: { nonce }}) {
 if (methodId === 'googlepayadyenv2') {
   return JSON.stringify({
                type: AdyenPaymentMethodType.GooglePay,
                googlePayToken: nonce,
                browser_info: getBrowserInfo(),
            });
  }

  return nonce;

}
```